### ISSUE TITLE
prepDE.py key errors with multiple sample directories

### ISSUE THREAD

**Issue by kotoroshinoto** [gpertea/stringtie#110]
I saw this was reported as a secondary concern in a closed issue, but as this wasn't the topic of that issue I wanted to report that I am having this issue with version 1.3.3

**Comment by kotoroshinoto** [comment]
the dictionary is initializing its keys off of a single sample,but it needs to add keys for other gene IDs when it encounters them. Ultimately that is the problem

**Comment by gpertea** [comment]
I disagree, if you follow the protocol as described (and if StringTie ran as it should), all the GTF files that prepDE is loading should have exactly the same sets of transcripts & genes (because these IDs are given in the `-G` input file, which should be the same for all samples). That's why loading the gene IDs and transcript IDs from the first sample should be enough.

If you are confident that all those GTFs were generated by StringTie 1.3.3 ran with the `-e`  option and with the same `-G` reference file, then I guess it is possible that there could still be a bug there that makes StringTie fail to output some of the input gene IDs. If you can find such an example I would be grateful if you can share it with us.  I recall we had such a bug with the `-e` option in an older version but I thought we stomped it out..

**Comment by kotoroshinoto** [comment]
thats odd because I'm pretty reasonably sure I used the same GTF with all my samples files.

I'll have to make sure i'm not missing a commandline argument somewhere.

**Comment by kotoroshinoto** [comment]
The GTF files have the command listed in them. It seems that indeed I did not have the -e option. My mistake.

**Comment by kotoroshinoto** [comment]
I understand that predicting transcripts will result in there being potentially different IDs. 

Is there no way to include a step of resolving & collapsing these so that they can be used with prepDE?

**Comment by gpertea** [comment]
Yes, that step of "resolving & collapsing" predicted transcripts is `stringtie --merge` and it is what you apparently *also* missed from the protocol (.. besides `-e`).

By the way, the output from just predicting transcripts (i.e. running stringtie without `-e`) is not just "potentially different IDs", it can be a lot different, and more importantly, all those with the assigned STRG prefix would have no meaning across samples *when they have the same ID!* (since those numeric IDs are only locally meaningful and are assigned internally for all potentially novel or incompletely assembled transcripts in a sample..).

**Comment by kotoroshinoto** [comment]
Thank you for that. Certainly I should have noticed it, but i'm on the right track now at least. (i'm exploring stringtie for the first time, and its been quite some time since I worked with cufflinks)

**Comment by kotoroshinoto** [comment]
Is there a way to ensure that the merge step uses the names/IDs from the GTF file preferentially, so that it only keeps the intenrally produced names for genes or transcripts that weren't in the GTF file?

**Comment by kotoroshinoto** [comment]
I only ask this because I noticed many cases where the genes had official symbols and yet were still using the string-tie IDs in the merged output

**Comment by gpertea** [comment]
Unfortunately it's not that easy, but I think it's worth taking another look at this as a future enhancement (advice/comments are welcome). 
Indeed, when multiple transcripts are merged or assembled by `stringtie --merge`, there will be new MSTRG-prefixed gene_ids assigned and the original gene ID information gets lost. It is difficult for StringTie to determine that the original gene_id information is worth keeping -- because some gene_ids may be completely bogus, as it's the case with some UCSC annotation files I've seen (where gene_id is just a copy of transcript_id). Now if you're talking about other attributes (like gene_name etc.), there could be a common ground there, but all those are optional features that StringTie cannot count on. This might be the job of a separate tool (e.g. something like gffcompare, though it doesn't do exactly what I describe here yet) that would allow the user to annotate a "merge" file with gene_name and other meaningful attributes taken from a reference annotation file, per user request. I'll have to think about this more.

**Comment by jokelley** [comment]
When I use the -e option, I am still getting several genes with the STRG prefix in the gene count matrix, why is this happening?
